### PR TITLE
Middleware cache and block at

### PR DESCRIPTION
**Pull request by ermalkaleci**

### CODE REVIEW

File: src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+use jsonrpsee::core::JsonValue;
+use lru::LruCache;
+use std::num::NonZeroUsize;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+pub struct Cache {
+    map: Arc<Mutex<LruCache<String, JsonValue>>>,
```
Comment (xlc) [inline src/cache.rs:8]: use RwLock

File: src/middleware/cache.rs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            return next(request).await;
+        }
+
+        let params_joined = request
+            .params
+            .iter()
+            .map(|x| x.to_string())
+            .collect::<Vec<_>>()
+            .join(",");
+        let key = format!("{}::[{}]", request.method, params_joined);
```
Comment (xlc) [inline src/middleware/cache.rs:34]: the key should be hash of params as params could be large
and using haser we don't need to join the paramers

File: src/cache.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    pub async fn get(&self, key: &String) -> Option<JsonValue> {
+        self.map.lock().await.get(key).cloned()
+    }
+
+    pub fn put(&self, key: String, value: JsonValue) {
+        let map = self.map.clone();
+        tokio::spawn(async move {
+            let _ = map.lock().await.put(key, value);
```
Comment (xlc) [inline src/cache.rs:25]: don't just ignore error. at least log it

File: src/middleware/inject_params.rs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            return next(request).await;
+        }
+
+        if let Some(param) = self.needs_to_inject(request.params.len()).await {
+            log::debug!("Injected param {} to method {}", &param, request.method);
+            request.params.push(param);
+            return next(request).await;
+        }
+
+        request.skip_caching = true;
```
Comment (xlc) [inline src/middleware/inject_params.rs:67]: this is unreachable right?

Comment (ermalkaleci) [reply src/middleware/inject_params.rs:67]: This is case number of params is not as expected

File: src/middleware/cache.rs (hunk 2 above)
Comment (ermalkaleci) [reply src/middleware/cache.rs:34]: Lru has it's own hasher but still we can hash it

File: src/cache.rs (hunk 3 above)
Comment (ermalkaleci) [reply src/cache.rs:25]: No error, ignoring old value

File: src/cache.rs (hunk 1 above)
Comment (ermalkaleci) [reply src/cache.rs:8]: don't see any benefit to using RwLock since `get` will update position so it required `write`

Comment (xlc) [reply src/cache.rs]: what do you mean? a cache hit will not need write

Comment (ermalkaleci) [reply src/cache.rs]: Lru cache removes the least used value so it updates on get

Comment (xlc) [reply src/cache.rs]: I see. In that case we need something that doesn't do that. It will completely kill the performance.

**Comment by xlc** [comment]
Some improvements are still required but can happen in later PR

File: src/cache.rs (hunk 1 above)
Comment (ermalkaleci) [reply src/cache.rs]: We can use peek which doesn't update anything but will lose the feature

**Pull request merged**